review

davidm-m opened the pull request.

davidm-m committed:
added firebase and testing rules

davidm-m committed:
Removed test; cannot place ships next to each other

davidm-m committed:
Added interfaces and database functionality

davidm-m committed:
fixing reading from database

davidm-m committed:
Moving close paren good god

davidm-m committed:
commented out reading from database

davidm-m committed:
removed all database funcitonality

davidm-m committed:
importing firebase

davidm-m committed:
Putting everything back in but it works this time hopefully

davidm-m committed:
Updating to Retry

davidm-m committed:
Removed @types/firebase

davidm-m committed:
Actually removed @types/firebase

davidm-m committed:
commented out export in index.d.ts for @types/firebase

davidm-m committed:
Re-added @types/firebase as a dependency

davidm-m committed:
Merge branch 'master' of https://git.heroku.com/afternoon-atoll-56925

davidm-m committed:
Separated getShipPositions into methods for each ship

davidm-m committed:
Uninstalled @types/firebase

firebase typings are included with firebase, no need for @types

davidm-m committed:
Testing adding to database

davidm-m committed:
Initialised variables

davidm-m committed:
experimental firebase change

davidm-m committed:
Getting heroku to unistall @types/firebase

davidm-m committed:
Same as before

davidm-m committed:
Check for collisions in matchId on firebase

davidm-m committed:
removed uninstall @types/firebase command from npm start

davidm-m committed:
First commit

davidm-m committed:
Updated backage.json to permanantly delete @types/firebase

davidm-m committed:
Merge branch 'master' of https://git.heroku.com/afternoon-atoll-56925

davidm-m committed:
Made collision checking actually work

davidm-m committed:
Removed uninstall command and now logs gamestate

davidm-m committed:
Added the uninstall command back in because heroku is dumb dumb dumb dumb dumb dumb

davidm-m committed:
Testing heroku writing to database

davidm-m committed:
Removed console logging gamestate

davidm-m committed:
First shot is now random

davidm-m committed:
Changed what is logged to database and randomshot now does not shoot places already shot

davidm-m committed:
Changed what is logged to database and randomshot now does not shoot places already shot

davidm-m committed:
Merge branch 'master' of https://git.heroku.com/afternoon-atoll-56925

# Conflicts:
#	src/my-bot.ts

davidm-m committed:
Fixed reading from database

davidm-m committed:
Removed erroneous statement

davidm-m committed:
Testing reading from database

davidm-m committed:
Trying new method of reading from database

davidm-m committed:
Issue wasn't with reading from database, it was dealing with the data

davidm-m committed:
Updated package.json to describe the bot rather than the template

davidm-m committed:
Testing why it still doesn't work

davidm-m committed:
Updated package.json because node is a square

davidm-m committed:
Added promise to getting database data

davidm-m committed:
My hands are typing words

davidm-m committed:
More debugging

davidm-m committed:
See last commit

davidm-m committed:
You know what it is

davidm-m committed:
Removing some logging

davidm-m committed:
Dealing with promises from firebase part 1

davidm-m committed:
Dealing with firebase promises part 2

davidm-m committed:
Part 3 of the messing with firebase promises trilogy

I'm expecting this to immediately throw an excpetion but a guy can hope

davidm-m committed:
Promises are working, just forgot to offset my random number generator

davidm-m committed:
Random shot generator is shooting the same place multiple times when it should not do that

Just a little logging to see what the issue may be

davidm-m committed:
MyShots is an array of Shots and I was chekcing if it contained a Position

Silly me

davidm-m committed:
More loggin for fixin

davidm-m committed:
Maybe directly comparing objects like that doesn't work

davidm-m committed:
Ship positions should now be randomly generated

davidm-m committed:
Error logging and seeing why the last commit is bugged

davidm-m committed:
Fixed random generation of places

davidm-m committed:
Actually fixed random generation of places

davidm-m committed:
Ships can no longer be next to each other

davidm-m committed:
Types are annoying

davidm-m committed:
Loggiiiiing

davidm-m committed:
More logging

davidm-m committed:
Fixed so generateshipsquares actually iterates

davidm-m committed:
generateAdjecentSquares works now

davidm-m committed:
Friendly reminder that types are annoying

davidm-m committed:
Logging

davidm-m committed:
Should actually detect collisions now

davidm-m committed:
On hitting a ship the targetter should continue hitting that ship until it is sunk

Expecting all kinds of cool errors from this one

davidm-m committed:
Set to intialise matches/id/hitmap

davidm-m committed:
Logging some stuff

davidm-m committed:
More logging and getting it to shoot randomly when not in hitmode

davidm-m committed:
Changed logging

davidm-m committed:
Added a catch for the promise I believe is throwing the exception

davidm-m committed:
Changed how hitmap works

davidm-m committed:
logging

davidm-m committed:
Possibly fixed getHitmap

davidm-m committed:
Removed unnecessary logging

davidm-m committed:
Looooooogggiiiiiiiinnnngggggg

davidm-m committed:
More logging

I'm trying to fix the targeting algorithm and lastHit doesn't seem to be created properly

davidm-m committed:
lastHit now works

Wew that was silly, need to remember to look at MyShots[i] not just MyShots!

davidm-m committed:
More logging to see why it always goes right

davidm-m committed:
More logging

davidm-m committed:
Didn't commit the right thing last time

davidm-m committed:
Hopefully fixed targeting

davidm-m committed:
Logging because I did not fix targeting

davidm-m committed:
Typescript is too afraid to log stuff after a return statement

davidm-m committed:
At this point I do not know what's going on

davidm-m committed:
Improved logging specifics

davidm-m committed:
[Object object]

davidm-m committed:
Possibly fixed targeting but if this works I will be sad

davidm-m committed:
Cut down on your promises to reduce headaches

davidm-m committed:
Scrapping an unnecessary promise worked, now logging to see why the targeting shot off the board

davidm-m committed:
Skipping problem should be fixed

davidm-m committed:
Stopped it from shooting in the same place forever

davidm-m committed:
Logging

davidm-m committed:
Fixed logs and forcing it to start where I know there's a problem

davidm-m committed:
Fixed shooting off board

davidm-m committed:
I think it might think boats are bigger than they are

davidm-m committed:
Removed boatsize stuff because it wasn't working

davidm-m committed:
Changed starting shot

davidm-m committed:
Changed to hit every other square with randomshot

davidm-m committed:
Got it to hit every other square, not every 4th

davidm-m committed:
Error catching

davidm-m committed:
Fixed trying to read an non-existent array

davidm-m committed:
Edited the ts file this time! Wow!

Last commit just increased the number ofloops needed to throw an exception when placing ships

davidm-m committed:
Logging

davidm-m committed:
Hopefully fixed trying to read array that doesn't exist

davidm-m committed:
Should not try to read an array that doesn't exist anymore, no matter when

davidm-m committed:
Empties the ship placement array when the loop runs

davidm-m committed:
Starts with a random shot again and fixed a bug in collision detecting for placing ships

davidm-m committed:
Fixed another bug in collision detection

davidm-m committed:
Fixed missing the last space on a boat if the space after had already been hit

As in, it wouldn't shoot at [x+1,y] if [x,y] is a hit and [x+2,y] is a miss

davidm-m committed:
Testing the previous commit

davidm-m committed:
More testing just to be sure

davidm-m committed:
Once more, with feeling

davidm-m committed:
It seems to work, removing test code

davidm-m committed:
Made generateadjacentsquares generate diagonals properly

davidm-m committed:
Fixed more array reading errors

davidm-m committed:
Logging to debug ship placement

davidm-m committed:
loggin'

davidm-m committed:
Changed logging and increased number of loops

davidm-m committed:
Moved the logging commands outside of the loops so it doesnt get printed quite so many times

davidm-m committed:
Investigating whether detectCollisions is sent the correct parameter

davidm-m committed:
Updated readme

davidm-m committed:
Changing logging

davidm-m committed:
Changed a little bit of how ship placement works

davidm-m committed:
Changed when the placement methods return

As far as I can tell, logically this shouldn't have any effect. However, the logs show that I must be getting something wrong

davidm-m committed:
Condensed the get ship placement methods into one method that now takes a size parameter, and simplified the code there

davidm-m committed:
Removed some unnecessary stuff and added boatsize checking back in

davidm-m committed:
Will no longer fire adjacent to a ship that has been sunk already

Hopefully

davidm-m committed:
Now won't pass hitmap to randomshot if hitmap doesn't exist

davidm-m committed:
generateAdjecentSquares handles being given an emtpy array

davidm-m committed:
generateAdjacentSquares now adds diagonals as it goes

davidm-m committed:
target is initialised to A1

RooftopTal commented on README.md:
```diff
@@ -1,14 +1,3 @@
-
-If you have done all the above successfully, you should see "Hello world!" printed to the console. If so, you're ready to start writing code!
\ No newline at end of file
+Worse than the rest, brokenbot is now playing battleships. Written in typescript.
```
Welp.

RooftopTal commented on package.json:
```diff
@@ -1,18 +1,19 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "start": "tsc -p . && node router.js"
+    "start": "npm uninstall @types/firebase --save && tsc -p . && node router.js"
```
Probably want to remove this now that you've fixed heroku :p

RooftopTal commented on router.ts:
```diff
@@ -13,19 +13,44 @@ export class Router {
+            } catch(err) {
+                console.log(err);
+                res.send(err);
+            }
```
Nice - I like this setup for your errors. Should have made it much easier for you to debug!

(Could commonise the log & send, so you're not repeating yourself, but it's a very minor thing!

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+import { ShipPlace } from './interfaces/shipPlace';
+import { Shot } from './interfaces/shot';
+
 export class MyBot {
```
Just going to throw it out there - this class was so large, github wouldn't load it by default!

I think you could easily be thinking about how you could cut this up into separate classes.

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+        authDomain: "brokenbot-battleships.firebaseapp.com",
+        databaseURL: "https://brokenbot-battleships.firebaseio.com",
+        storageBucket: "brokenbot-battleships.appspot.com"
+    };
```
Do you need this one and the one in the router?

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+            counter++;
+            if (counter > 10000) {
+                throw new Error("Infinite loop when constructing bot");
+            }
```
Nice safety valve.

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+            if (counter > 10000) {
+                throw new Error("Infinite loop when constructing bot");
+            }
+        } while (exists)
```
This could easily be pulled out into it's own method - something like  `findExistingSnapshot`, or something.

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+                carrier: true
+            },
+            hitmap: []
+        });
```
To be honest, I would pull out all of these firebase methods onto their own class - that way another dev knows that that's the interface with firebase

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+                submarine: true,
+                destroyer: true,
+                battleship: true,
+                carrier: true
```
cute!

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+                continue;
+            }
+            done = true;
+        } while ((!done) && (counter < 10000))
```
Similarly, this could be separated into `getShipPlaces`

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
     }
 
-    private getNextTarget(position) {
+    public selectTarget(gamestate): Promise<Position> {
```
These should really be typed!

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
-    private getNextTarget(position) {
+    public selectTarget(gamestate): Promise<Position> {
+        return new Promise((resolve, reject) => {
+            firebase.database().ref('matches/' + this.matchId.toString()).once('value').then((snapshot) => { 
```
I am impressed at the use of firebase - you've really engaged with it in some quite cool ways.

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+                }
+                return resolve(target);
+            })
+            .catch((err) => {console.log(err)});
```
Neat, but I'm a little wary of this simply logging and hiding the error. What errors are you expecting - firebase ones? Wouldn't it be better to rethrow them so that your error handling on the router catches them?

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+    private authenticate(): void {
+        firebase.auth().signInWithEmailAndPassword(
+            "david.may-miller@softwire.com",
+            "securePassword3"
```
Woop.

(Obviously, in general, hardcoding passwords in git history is a bad thing, but it's very practical (the DMP project also has this atm)

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+        return { StartingSquare: { Row: "A", Column: -1 }, EndingSquare : { Row: "A", Column: -1 } };
+    }
+
+    private detectCollision(shipPlaces: ShipPlace[]): boolean { //apart from the last one, none of these for loops should be more than 5 iterations so it's not as bad as it looks
```
I like your naming - I don't actually think that your comment is needed, really. While there's a lot of nesting, this is cut down small enough that it's easy to pick up what it's doing.

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+            squares.push({ Row: String.fromCharCode(ship[i].Row.charCodeAt(0) + 1), Column: ship[i].Column + 1 });
+            squares.push({ Row: String.fromCharCode(ship[i].Row.charCodeAt(0) - 1), Column: ship[i].Column + 1 });
+            squares.push({ Row: String.fromCharCode(ship[i].Row.charCodeAt(0) + 1), Column: ship[i].Column - 1 });
+            squares.push({ Row: String.fromCharCode(ship[i].Row.charCodeAt(0) - 1), Column: ship[i].Column - 1 });
```
This would be much nicer if this had a proper type - e.g.

`squares.push (new RowColumn(ship[i].Row, ship[i].Column + 1));`

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+    }
+
+    private randomShot(gamestate, hits: Position[] = []): Position {
+        const useless: Position[] = this.generateAdjacentSquares(hits);
```
Having a const called useless is probably not the best! Find a more obvious name for it!

RooftopTal commented on src/my-bot.ts:
```diff
@@ -1,38 +1,423 @@
+        return position
+    }
+
+    private track(gamestate, snapCopy): Position {
```
This is another really long method. I think you could almost:

* Make this a class called `TrackState` or similar
* Set up your variables as private vars
* Cut it into a bunch of methods (`setup`, `getDirection`, etc.)

...and all of these would be a lot easier to follow.

RooftopTal reviewed (changes requested):
In general, I'm pretty happy - you're writing well named, well typed code, and you've really gone deep on all of the tools available to make a bot (I'm particularly impressed with how deep you've gone in to firebase!).

I think that you do tend towards large classes/large methods, however. Your MyBot class is probably doing the work of at least 3-4 classes, and some of the methods could be cut up a lot more than they are. Something to pay attention to in your first few cards.

davidm-m committed:
Made some changes based on review

Removed unnecessary config object from router.ts; moved firebase handling to its own class; created a gamestate interface for typing; generally cleaned up a little

davidm-m committed:
Fixed finding a free matchId

davidm-m committed:
Fiddled with promises to make sure matchId is generated before it needs to be used

davidm-m committed:
Removed uninstall @types/firebase command from npm start

I don't think this will work. Also hopefully fixed last commit

davidm-m committed:
This should have been the last commit really

davidm-m committed:
Fixed problem with findfreeid, was suing this.matchId instead of matchId!

Silly!

davidm-m committed:
Fixed problem with findfreeid, was setting this.matchId instead of matchId!

Silly!

davidm-m committed:
Merge branch 'master' of https://github.com/davidm-m/basic-battleship-bot-typescript-heroku
